### PR TITLE
Chainlink BNB Price Oracle

### PR DESCRIPTION
**Pull request by 0xroll**
creating for chainlink price feed for bnbchain

similar to this : https://github.com/duneanalytics/spellbook/pull/2048

Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

File: models/chainlink/bnb/chainlink_bnb_price_feeds_hourly.sql [hunk 1]
```diff
@@ -0,0 +1,124 @@
+{{ config(
+    alias = 'price_feeds_hourly',
+    partition_by = ['block_date'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['blockchain', 'hour', 'proxy_address', 'aggregator_address', 'underlying_token_address'],
```
Comment (Hosuke) [inline models/chainlink/bnb/chainlink_bnb_price_feeds_hourly.sql:7]: Questions:
1. `partition_by = ['block_date']` or `partition_by = ['hour']`
2. if `partition_by = ['block_date']`, should `'block_date'` go into unique_key?

**Review by Hosuke — APPROVED** [review]
To whom it may concern, I have reviewed this PR and here is the summary:

- Metadata and naming convention looks good ✅
- Config blocks match uniqueness tests in schema files ✅
- Tests and seeds added for the new models ✅
- Incremental filters added when joining large tables ✅
- Contract addresses are lowercased ✅
- Inner join to large tables(bnb.logs) as possible ✅

File: models/chainlink/bnb/chainlink_bnb_price_feeds.sql [hunk 2]
```diff
@@ -4,7 +4,7 @@
     materialized = 'incremental',
     file_format = 'delta',
     incremental_strategy = 'merge',
-    unique_key = ['blockchain', 'block_time', 'proxy_address', 'oracle_price', 'aggregator_address', 'underlying_token_address'],
+    unique_key = ['blockchain', 'block_number', 'proxy_address', 'aggregator_address', 'underlying_token_address'],
```
Comment (Hosuke) [inline models/chainlink/bnb/chainlink_bnb_price_feeds.sql:7]: Hi @0xroll , in my last commit I change the unique_key list to 5 elements, please have a check if this still works.

**Comment by Hosuke** [comment]
Hi @MSilb7, for your information, @0xroll and I make some slight changes on the chainlink OP models in the last commit:
- simplify the unique key used in the OP model
- change `block_time` to `block_number` in the unique key (for more stable joining and testing)
- add unit_test and check_seed tests
- add a few seeds

If you have any suggestion on these changes please let us know, thank you.

**Comment by MSilb7** [comment]
> Hi @MSilb7, for your information, @0xroll and I made some slight changes on the chainlink OP models in the last commit:
> 
> * simplify the unique key used in the OP model
> * change `block_time` to `block_number` in the unique key (for more stable joining and testing)
> * add unit_test and check_seed tests
> * add a few seeds
> 
> If you have any suggestion on these changes please let us know, thank you.

Awesome, sounds good!

File: models/chainlink/optimism/chainlink_optimism_price_feeds.sql [hunk 3]
```diff
@@ -4,11 +4,11 @@
     materialized = 'incremental',
     file_format = 'delta',
     incremental_strategy = 'merge',
-    unique_key = ['blockchain', 'block_time', 'proxy_address', 'oracle_price', 'aggregator_address', 'underlying_token_address'],
+    unique_key = ['blockchain', 'block_number', 'proxy_address'],
```
Comment (MSilb7) [inline models/chainlink/optimism/chainlink_optimism_price_feeds.sql:7]: I wonder if we'd need 'aggregator_address' here? Not sure if there's ever a case where a proxy is used multiple times.

Comment (0xroll) [reply models/chainlink/optimism/chainlink_optimism_price_feeds.sql]: hey @MSilb7 ! yes we can remove the aggregator_address for the unique_key, but will have to add in the underlying_token_address as there can be >1 token using the same feed

File: seeds/chainlink/chainlink_get_price_seed.csv [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+polygon,2022-11-23 03:10,35950884,GBP / USD,1.18855,0x099a2540848573e94fb1ca0fa420b00acbbc845a,0x3f7f90e0f782e325401f6323ba93e717f519f382,0x767058f11800fba6a682e73a6e79ec5eb74fac8c
+polygon,2022-11-23 03:05,35950735,AUD / USD,0.66465,0x062df9c4efd2030e243ffcc398b652e8b8f95c6f,0x0a9823c5cd84099fde8566a1adf0f2bb41cc6e7d,0xcb7f1ef7246d1497b985f7fc45a1a31f04346133
+polygon,2022-11-23 03:11,35950885,CAD / USD,0.747468,0xaca44abb8b04d07d883202f99fa5e3c53ed57fb5,0x88b79bfce730bbb74f23ab8940b37b86859caa2e,0x8ca194a3b22077359b5732de53373d4afc11dee3
+polygon,2022-11-17 02:07,35701896,SGD / USD,0.72890016,0x8ce3cac0e6635ce04783709ca3cc4f5fc5304299,0x45ede0ea5cbbe380c663c7c3015cc7c986669fec,0xa926db7a4cc0cb1736d5ac60495ca8eb7214b503
+polygon,2021-07-30 10:55,17423257,CHF / USD,1.104948,0xc76f762cedf0f78a439727861628e0fdfe1e70c2,0x8123beacb5bca3afa0c9ff71b28549d58cec8176,0xbd1463f02f61676d53fd183c2b19282bff93d099
+polygon,2021-11-18 07:15,21507237,NZD / USD,0.70375,0xa302a0b8a499fd0f00449df0a490dede21105955,0xe63032a70f6eb617970829fbfa365d7c44bdbbbf,0x08e6d1f0c4877ef2993ad733fc6f1d022d0e9dbf
```
Comment (Hosuke) [inline seeds/chainlink/chainlink_get_price_seed.csv:12]: Hi @0xroll, please have a look on this seed, as this record may not be available in the [query result](https://dune.com/queries/1707702?d=1)

**Comment by Hosuke** [comment]
Hi @jeff-dude , it seems this PR is ready for the final review.

This PR contains:
- new chainlink models for bnb chain
- enhancements for other 2 chainlink models on polygon and optimism chain
- seeds and tests for these 3 chainlink models

The further changes are legitimate to me since my last review.

File: models/chainlink/bnb/chainlink_bnb_price_feeds_hourly.sql (hunk 1 above)
Comment (jeff-dude) [reply models/chainlink/bnb/chainlink_bnb_price_feeds_hourly.sql]: `block_date` is good for partition -- it does not need to be in unique keys just for the partitioning.

**Comment by jeff-dude** [comment]
thank you @0xroll and @Hosuke -- this looks good to me!

**Comment by jeff-dude** [comment]
@belen-pruvost this contains a few new view models that may have syntax for you to validate -- can you quickly look through each model which isn't explicitly configured as incremental and see if any views need syntax change?

**Review by jeff-dude — APPROVED** [review]

**Comment by belen-pruvost** [comment]
Hey @jeff-dude - I missed your comment, sorry. I've opened a [PR](https://github.com/duneanalytics/spellbook/pull/2275) with the syntax changes needed for compatibility.

**Pull request merged**